File: admob/InAppPurchaseExample/app/src/main/java/com/google/android/gms/example/inapppurchaseexample/MainActivity.java
/*
 * Copyright (C) 2013 Google, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.android.gms.example.inapppurchaseexample;

import android.app.Activity;
import android.os.Bundle;
import android.support.v7.app.ActionBarActivity;
import android.support.v7.app.AppCompatActivity;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import com.google.android.gms.ads.AdListener;
import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.InterstitialAd;
import com.google.android.gms.ads.MobileAds;
import com.google.android.gms.ads.purchase.InAppPurchaseResult;
import com.google.android.gms.ads.purchase.PlayStorePurchaseListener;

import java.util.HashSet;
import java.util.Set;

/**
 * A class that demonstrates how to use AdMob in-app purchase ads.
 */
public class MainActivity extends AppCompatActivity implements PlayStorePurchaseListener {

    /**
     * The public key for your Android application, which can be found in the Android console.
     *
     * This parameter is optional, but it is recommended to include in a real application that uses
     * billing. Since your app defined on Android is for a different application package than this
     * sample, this parameter can be left as is for the purposes of running the sample.
     */
    private static final String PUBLIC_KEY = "<YOUR_PUBLIC_KEY>";

    /**
     * Your ad unit ID for in-app purchase ads.
     */
    private static final String AD_UNIT_ID = "/6499/example/iap";

    /**
     * Your AdMob Application ID.
     */
    private static final String APP_ID = "ca-app-pub-3940256099942544~3347511713";

    /**
     * The key for the billing response code in the Intent extras.
     */
    public static final String BILLING_RESPONSE_CODE_KEY = "RESPONSE_CODE";

    /**
     * The response code for a successful purchase.
     */
    public static final int BILLING_RESPONSE_RESULT_OK = 0;

    /**
     * The set of SKUs that this app can handle.
     */
    private Set<String> mValidSkus;

    /**
     * The interstitial used to load in-app purchase ads.
     */
    private InterstitialAd mInterstitial;

    /**
     * The amount of goods owned.
     */
    private int mGoods;

    /**
     * The view indicating the amount of goods owned.
     */
    private TextView mGoodsTextView;

    /**
     * The load interstitial button.
     */
    private Button mLoadInterstitialButton;

    /**
     * The show interstitial button.
     */
    private Button mShowInterstitialButton;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);


        // Initialize the Mobile Ads SDK.
        MobileAds.initialize(this, APP_ID);

        mGoods = 0;
        mGoodsTextView = (TextView) findViewById(R.id.tv_goods);
        updateGoods();

        // SKUs that your app supports.
        mValidSkus = new HashSet<String>();
        mValidSkus.add("android.test.purchased");

        mInterstitial = new InterstitialAd(this);
        mInterstitial.setAdUnitId(AD_UNIT_ID);
        mInterstitial.setPlayStorePurchaseParams(this, PUBLIC_KEY);
        mInterstitial.setAdListener(new AdListener() {
            @Override
            public void onAdLoaded() {
                Toast.makeText(MainActivity.this, "Ad loaded", Toast.LENGTH_SHORT).show();
                mShowInterstitialButton.setEnabled(true);
            }

            @Override
            public void onAdFailedToLoad(int errorCode) {
                Toast.makeText(MainActivity.this, "Ad failed to load", Toast.LENGTH_SHORT).show();
                mLoadInterstitialButton.setEnabled(true);
            }

            @Override
            public void onAdOpened() {
                mShowInterstitialButton.setEnabled(false);
            }

            @Override
            public void onAdClosed() {
                mLoadInterstitialButton.setEnabled(true);
            }
        });

        mLoadInterstitialButton = (Button) findViewById(R.id.btn_load_ad);
        mLoadInterstitialButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View unusedView) {
                mLoadInterstitialButton.setEnabled(false);
                AdRequest request = new AdRequest.Builder().build();
                mInterstitial.loadAd(request);
            }
        });

        mShowInterstitialButton = (Button) findViewById(R.id.btn_show_ad);
        mShowInterstitialButton.setEnabled(false);
        mShowInterstitialButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View unusedView) {
                if (mInterstitial.isLoaded()) {
                    mInterstitial.show();
                }
            }
        });
    }

    /**
     * Updates the number of goods owned.
     */
    private void updateGoods() {
        mGoodsTextView.setText(String.format("Goods: %d", mGoods));
    }

    @Override
    public boolean isValidPurchase(String sku) {
        boolean validPurchase = mValidSkus.contains(sku);
        Toast.makeText(this,
                String.format("isValidPurchase for product %s: %b", sku, validPurchase),
                Toast.LENGTH_SHORT).show();
        return validPurchase;
    }

    @Override
    public void onInAppPurchaseFinished(InAppPurchaseResult result) {
        String sku = result.getProductId();
        if (result.getResultCode() == Activity.RESULT_OK) {
            Toast.makeText(this,
                    String.format("Successful purchase for product %s", sku),
                    Toast.LENGTH_SHORT).show();
            // Reward user with goods.
            mGoods += 5;
            updateGoods();
            // Remember to call finishPurchase() to consume the purchase.
            result.finishPurchase();
        } else {
            Toast.makeText(this,
                    String.format("Failed to purchase product %s", sku),
                    Toast.LENGTH_SHORT).show();
        }
    }
}

File: kotlin/admob/BannerExample/app/src/main/java/com/google/android/gms/example/bannerexample/MainActivity.kt
/*
 * Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.android.gms.example.bannerexample

import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.PopupMenu
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import com.google.android.gms.ads.AdListener
import com.google.android.gms.ads.AdRequest
import com.google.android.gms.ads.AdSize
import com.google.android.gms.ads.AdView
import com.google.android.gms.ads.LoadAdError
import com.google.android.gms.ads.MobileAds
import com.google.android.gms.ads.RequestConfiguration
import com.google.android.gms.example.bannerexample.databinding.ActivityMainBinding
import java.util.concurrent.atomic.AtomicBoolean
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

/** Main Activity. Inflates main activity xml and child fragments. */
class MainActivity : AppCompatActivity() {

  private val isMobileAdsInitializeCalled = AtomicBoolean(false)
  private var adView: AdView? = null
  private lateinit var binding: ActivityMainBinding
  private lateinit var googleMobileAdsConsentManager: GoogleMobileAdsConsentManager

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    binding = ActivityMainBinding.inflate(layoutInflater)
    setContentView(binding.root)
    setSupportActionBar(binding.toolBar)

    ViewCompat.setOnApplyWindowInsetsListener(binding.root) { view, insets ->
      val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
      view.setPadding(systemBars.left, 0, systemBars.right, systemBars.bottom)
      insets
    }

    // Log the Mobile Ads SDK version.
    Log.d(TAG, "Google Mobile Ads SDK Version: " + MobileAds.getVersion())

    googleMobileAdsConsentManager = GoogleMobileAdsConsentManager.getInstance(applicationContext)
    googleMobileAdsConsentManager.gatherConsent(this) { error ->
      if (error != null) {
        // Consent not obtained in current session.
        Log.d(TAG, "${error.errorCode}: ${error.message}")
      }

      if (googleMobileAdsConsentManager.canRequestAds) {
        initializeMobileAdsSdk()
      }

      if (googleMobileAdsConsentManager.isPrivacyOptionsRequired) {
        // Regenerate the options menu to include a privacy setting.
        invalidateOptionsMenu()
      }
    }

    // This sample attempts to load ads using consent obtained in the previous session.
    if (googleMobileAdsConsentManager.canRequestAds) {
      initializeMobileAdsSdk()
    }
  }

  /** Called when leaving the activity. */
  public override fun onPause() {
    adView?.pause()
    super.onPause()
  }

  /** Called when returning to the activity. */
  public override fun onResume() {
    super.onResume()
    adView?.resume()
  }

  /** Called before the activity is destroyed. */
  public override fun onDestroy() {
    adView?.destroy()
    super.onDestroy()
  }

  override fun onCreateOptionsMenu(menu: Menu?): Boolean {
    menuInflater.inflate(R.menu.action_menu, menu)
    return super.onCreateOptionsMenu(menu)
  }

  override fun onOptionsItemSelected(item: MenuItem): Boolean {
    val menuItemView = findViewById<View>(item.itemId)
    val activity = this
    PopupMenu(this, menuItemView).apply {
      menuInflater.inflate(R.menu.popup_menu, menu)
      menu
        .findItem(R.id.privacy_settings)
        .setVisible(googleMobileAdsConsentManager.isPrivacyOptionsRequired)
      show()
      setOnMenuItemClickListener { popupMenuItem ->
        when (popupMenuItem.itemId) {
          R.id.privacy_settings -> {
            // Handle changes to user consent.
            googleMobileAdsConsentManager.showPrivacyOptionsForm(activity) { formError ->
              if (formError != null) {
                Toast.makeText(activity, formError.message, Toast.LENGTH_SHORT).show()
              }
            }
            true
          }
          R.id.ad_inspector -> {
            MobileAds.openAdInspector(activity) { error ->
              // Error will be non-null if ad inspector closed due to an error.
              error?.let { Toast.makeText(activity, it.message, Toast.LENGTH_SHORT).show() }
            }
            true
          }
          else -> false
        }
      }
      return super.onOptionsItemSelected(item)
    }
  }

  private fun loadBanner() {
    // [START create_ad_view]
    // Create a new ad view.
    val adView = AdView(this)
    adView.adUnitId = AD_UNIT_ID
    // [START set_ad_size]
    // Request a large anchored adaptive banner with a width of 360.
    adView.setAdSize(AdSize.getLargeAnchoredAdaptiveBannerAdSize(this, 360))
    // [END set_ad_size]
    this.adView = adView

    // Replace ad container with new ad view.
    binding.adViewContainer.removeAllViews()
    binding.adViewContainer.addView(adView)
    // [END create_ad_view]

    // Listen for ad events.
    adView.adListener =
      object : AdListener() {
        override fun onAdLoaded() {
          // Called when an ad is loaded.
          Log.d(TAG, "Ad loaded.")
        }

        override fun onAdFailedToLoad(error: LoadAdError) {
          // Called when an ad request failed.
          Log.i(TAG, "Ad failed to load: ${error.message}")
        }

        override fun onAdOpened() {
          // Called when an ad opens an overlay that covers the screen.
          Log.d(TAG, "Ad opened.")
        }

        override fun onAdClicked() {
          // Called when a click is recorded for an ad.
          Log.d(TAG, "Ad clicked.")
        }

        override fun onAdImpression() {
          // Called when an impression is recorded for an ad.
          Log.d(TAG, "Ad recorded an impression.")
        }

        override fun onAdClosed() {
          // Called when the user is about to return to the application
          // after tapping on an ad.
          Log.d(TAG, "Ad closed.")
        }
      }

    // [START load_ad]
    val adRequest = AdRequest.Builder().build()
    adView.loadAd(adRequest)
    // [END load_ad]
  }

  private fun initializeMobileAdsSdk() {
    if (isMobileAdsInitializeCalled.getAndSet(true)) {
      return
    }

    // Set your test devices.
    MobileAds.setRequestConfiguration(
      RequestConfiguration.Builder().setTestDeviceIds(listOf(TEST_DEVICE_HASHED_ID)).build()
    )

    // [START initialize_sdk]
    CoroutineScope(Dispatchers.IO).launch {
      // Initialize the Google Mobile Ads SDK on a background thread.
      MobileAds.initialize(this@MainActivity) {}
      // [START_EXCLUDE silent]
      runOnUiThread {
        // Load an ad on the main thread.
        loadBanner()
      }
      // [END_EXCLUDE]
    }
    // [END initialize_sdk]
  }

  companion object {
    // This is an ad unit ID for a test ad. Replace with your own banner ad unit ID.
    private const val AD_UNIT_ID = "ca-app-pub-3940256099942544/9214589741"
    private const val TAG = "MainActivity"

    // Check your logcat output for the test device hashed ID e.g.
    // "Use RequestConfiguration.Builder().setTestDeviceIds(Arrays.asList("ABCDEF012345"))
    // to get test ads on this device" or
    // "Use new ConsentDebugSettings.Builder().addTestDeviceHashedId("ABCDEF012345") to set this as
    // a debug device".
    const val TEST_DEVICE_HASHED_ID = "ABCDEF012345"
  }
}
